NXP folder architecture improvement

chapongatien opened the pull request:
**Description**:

Improving NXP matter folder architecture by moving NXP SDK support files in a dedicated submodule. 
The aim of this folder architecture rework is to reduce the number of NXP SDK files that are currently stored in the connectedhome repo by moving them in a dedicated “nxp_matter_support” repo that would be located in “third_party/nxp/nxp_matter_support”.

**Main changes summary:**

-	K32w0, k32w1 and rw61x files that are highly linked to NXP SDK have been moved to the nxp_matter_support submodule.
-	GN build files have been updated to reference new paths
-	Adding a new script called update_nxp_sdk.py allowing to ease initialization of recommended github NXP SDK(s) for platform that support it (k32w0 documentation has been updated) .

chapongatien committed:
[NXP][script] Adding an update nxp sdk script allowing to initialize all nxp SDK with west for all platform

Signed-off-by: Gatien Chapon <gatien.chapon@nxp.com>

chapongatien committed:
[NXP][doc][k32w0] Updating nxp sdk update steps to use new update_nxp_sdk.py script

+ updating environement setup documentation

Signed-off-by: Gatien Chapon <gatien.chapon@nxp.com>

chapongatien committed:
[NXP] Adding nxp_matter_support submodule

Adding a reference to the nxp_matter_support repo that would be used to get NXP SDK support files.

Signed-off-by: Gatien Chapon <gatien.chapon@nxp.com>

chapongatien committed:
[NXP] Moving NXP Matter SDK support files to a dedicated nxp_matter_support repo

Signed-off-by: Gatien Chapon <gatien.chapon@nxp.com>

chapongatien committed:
[NXP][k32w0] Updating default sdk path to nxp_matter_support

Signed-off-by: Gatien Chapon <gatien.chapon@nxp.com>

restyled-commits committed:
Restyled by gn

restyled-commits committed:
Restyled by prettier-markdown

andy31415 commented on build_overrides/nxp_sdk.gni:
```diff
@@ -14,5 +14,9 @@
+  nxp_sdk_matter_support_root = "//third_party/nxp/nxp_matter_support"
+}
+
+declare_args() {
```
could we merge this in a single declare args? saves some vertical space.

chapongatien replied:
@andy31415  I got a GN error which recommend to separate them in 2 declare_args (as below):

2024-06-20 09:40:44 INFO    ERROR at //build_overrides/nxp_sdk.gni:19:27: Reading a variable defined in the same declare_args() call.
2024-06-20 09:40:44 INFO    
2024-06-20 09:40:44 INFO    If you need to set the value of one arg based on another, put
2024-06-20 09:40:44 INFO    them in two separate declare_args() calls, one after the other.

Do you have other recommendation to fix this error ?

andy31415 commented on examples/build_overrides/nxp_sdk.gni:
```diff
@@ -14,5 +14,10 @@
+      "//third_party/connectedhomeip/third_party/nxp/nxp_matter_support"
+}
+
+declare_args() {
```
same here: could we merge into one declare args?

andy31415 commented on scripts/setup/nxp/update_nxp_sdk.py:
```diff
@@ -0,0 +1,96 @@
+    git_clean_command = "git reset --hard && git clean -xdf"
+    command = ['west', 'forall', '-c', git_clean_command, '-a']
+    subprocess.run(command, cwd=sdk_target_location, check=True)
+    # Retrying an update
```
why is this a "retry"?  It would be nice for the code to explain why things are done - I can see it does an update_platform_sdk_version, but not as clear why.

What does `False` mean?

chapongatien replied:
I agree, it is not very clear and the location of the retry is not correct. I re-worked this part my latest commit and moved this "retry" mechanism in a better place.  Please let me know if you are ok.

chapongatien committed:
[NXP][script] Updating update_nxp_sdk script to clean force update processing

Signed-off-by: Gatien Chapon <gatien.chapon@nxp.com>

andy31415 commented on scripts/setup/nxp/update_nxp_sdk.py:
```diff
@@ -0,0 +1,98 @@
+
+# Array listing all SDK and default targeted path for NXP SDK storage
+ALL_PLATFORM_SDK = [
+    {'plat_name': 'k32w0', 'plat_sdk_manifest_path': 'third_party/nxp/nxp_matter_support/github_sdk/k32w0'},
```
Consider replacing stringly-typed data with something like `@dataclass` or similar. 

a `[p.name for p in ALL_PLATFORM_SDK]` is more readable than the existing `list(map(lambda p: p['plat_name'], ALL_PLATFORM_SDK))`

chapongatien replied:
I did an update in the last commit to use dataclass. Thanks for your recommendation.

andy31415 commented on scripts/setup/nxp/update_nxp_sdk.py:
```diff
@@ -0,0 +1,98 @@
+    args = parser.parse_args()
+
+    for current_plat in args.platform:
+        plat_path = list(map(lambda res: res['plat_sdk_manifest_path'], filter(
```
this should probably read like:

```py
plat_path = [p.manifest_path for p in ALL_PLATFORM_SDK if p.name == current_plat][0]
```

- use dataclasses
- using `[]` seems more readable than list/map/lambda

andy31415 commented on scripts/setup/nxp/update_nxp_sdk.py:
```diff
@@ -0,0 +1,98 @@
+
+def init_platform_sdk_version(sdk_target_location, force):
+    print("Init SDK in: " + sdk_target_location)
+    if not force and os.path.exists(os.path.join(sdk_target_location, '.west')):
+        print("Error SDK is already initialized, use --force to force init")
+        print("WARNING -- All local SDK modification(s) will be lost")
+        sys.exit(1)
+    command = ['rm', '-rf', '.west']
+    subprocess.run(command, cwd=sdk_target_location, check=True)
```
```suggestion
    west_path = os.path.join(sdk_target_location, '.west')
    if  os.path.exists(west_path):
        if not force:
            print("Error SDK is already initialized, use --force to force init")
            print("WARNING -- All local SDK modification(s) will be lost")
            sys.exit(1)
        command = ['rm', '-rf', west_path]
        subprocess.run(command, check=True)
```

- seems to make it clearer and avoid running the remove if path does not exist.
- makes it clear you are deleting the exact path that is being checked for
- try to avoid having subprocess depend on cwd if possible

andy31415 commented on scripts/setup/nxp/update_nxp_sdk.py:
```diff
@@ -0,0 +1,98 @@
+            clean_sdk_local_changes(sdk_target_location)
+            subprocess.run(command, cwd=sdk_target_location, check=True)
+        else:
+            print(exception)
```
consider using python logging throughout. Then you can use `logging.exception` to report the exception with a full stack trace and info. otherwise probably traceback would be useful.

chapongatien replied:
Yes agree, I switched to python logging in the last commit.

chapongatien committed:
Updating nxp_sdk update script after code review

- dataclass is now used
- subprocess dependencies on cwd removed with possible (for west update and west forall cwd is still required)
- python logging module is now used

Signed-off-by: Gatien Chapon <gatien.chapon@nxp.com>
